File: pyserver/draw.py
"""Server-side draw execution and persistence of the running draw state."""

import secrets
from datetime import datetime, timezone
from typing import Any, Dict, List, Sequence

from .paths import DRAW_STATE_PATH
from .store import read_json, write_json
from .tickets import load_ticket_source

EMPTY_STATE: Dict[str, Any] = {"winners": [], "startedAt": None, "updatedAt": None}


def _now() -> str:
    return datetime.now(timezone.utc).isoformat()


def load_draw_state() -> Dict[str, Any]:
    state = read_json(DRAW_STATE_PATH, EMPTY_STATE) or EMPTY_STATE
    winners = state.get("winners")
    return {
        "winners": winners if isinstance(winners, list) else [],
        "startedAt": state.get("startedAt"),
        "updatedAt": state.get("updatedAt"),
    }


def save_draw_state(state: Dict[str, Any]) -> None:
    write_json(DRAW_STATE_PATH, {**state, "updatedAt": _now()})


def read_tickets(app_settings: Dict[str, Any]) -> Dict[str, Any]:
    return load_ticket_source(app_settings["data"], app_settings["data"]["duplicatePolicy"])


def remaining_pool(
    tickets: Sequence[Dict[str, Any]], winners: Sequence[Dict[str, Any]], identifier: str
) -> List[Dict[str, Any]]:
    drawn = {str((winner.get("record") or {}).get(identifier, "")).lower() for winner in winners}
    return [ticket for ticket in tickets if str(ticket.get(identifier, "")).lower() not in drawn]


def build_stats(
    app_settings: Dict[str, Any], tickets: Sequence[Dict[str, Any]], winners: Sequence[Dict[str, Any]]
) -> Dict[str, Any]:
    total_prizes = app_settings["totalPrizes"]
    remaining_tickets = len(remaining_pool(tickets, winners, app_settings["data"]["identifier"]))
    return {
        "totalPrizes": total_prizes,
        "winnersCount": len(winners),
        "remainingPrizes": max(0, total_prizes - len(winners)),
        "totalTickets": len(tickets),
        "remainingTickets": remaining_tickets,
        "isComplete": len(winners) >= total_prizes or remaining_tickets == 0,
    }


def draw_winner(app_settings: Dict[str, Any]) -> Dict[str, Any]:
    """Draw once and persist. Exhausted prizes are a result, not an exception."""
    tickets = read_tickets(app_settings)["tickets"]
    state = load_draw_state()

    if not tickets:
        return {"ok": False, "reason": "no-tickets", "message": "No participants have been uploaded yet."}

    if len(state["winners"]) >= app_settings["totalPrizes"]:
        return {"ok": False, "reason": "prizes-exhausted", "message": "All prizes have already been awarded."}

    pool = remaining_pool(tickets, state["winners"], app_settings["data"]["identifier"])
    if not pool:
        return {"ok": False, "reason": "pool-empty", "message": "Every ticket has already been drawn."}

    # secrets.randbelow is uniform and cryptographically sound, which matters
    # for a draw whose fairness has to be defensible.
    selected = pool[secrets.randbelow(len(pool))]
    winner = {"prizeNumber": len(state["winners"]) + 1, "drawnAt": _now(), "record": selected}

    next_state = {
        **state,
        "startedAt": state["startedAt"] or winner["drawnAt"],
        "winners": [*state["winners"], winner],
    }
    save_draw_state(next_state)

    return {"ok": True, "winner": winner, "stats": build_stats(app_settings, tickets, next_state["winners"])}


def undo_last_winner(app_settings: Dict[str, Any]) -> Dict[str, Any]:
    state = load_draw_state()
    if not state["winners"]:
        return {"ok": False, "reason": "nothing-to-undo", "message": "There are no draws to undo."}

    removed = state["winners"][-1]
    next_state = {**state, "winners": state["winners"][:-1]}
    save_draw_state(next_state)

    tickets = read_tickets(app_settings)["tickets"]
    return {"ok": True, "removed": removed, "stats": build_stats(app_settings, tickets, next_state["winners"])}


def reset_draw(app_settings: Dict[str, Any]) -> Dict[str, Any]:
    save_draw_state({"winners": [], "startedAt": None})
    return {"ok": True, "stats": build_stats(app_settings, read_tickets(app_settings)["tickets"], [])}

File: server/draw.js
'use strict';

const crypto = require('crypto');

const { PATHS } = require('./paths');
const { readJson, writeJson } = require('./store');
const { loadTicketSource } = require('./tickets');

const EMPTY_STATE = Object.freeze({ winners: [], startedAt: null, updatedAt: null });

function loadDrawState() {
  const state = readJson(PATHS.drawState, EMPTY_STATE) || EMPTY_STATE;
  return {
    winners: Array.isArray(state.winners) ? state.winners : [],
    startedAt: state.startedAt || null,
    updatedAt: state.updatedAt || null,
  };
}

function saveDrawState(state) {
  writeJson(PATHS.drawState, { ...state, updatedAt: new Date().toISOString() });
}

function readTickets(appSettings) {
  return loadTicketSource(appSettings.data, appSettings.data.duplicatePolicy);
}

function remainingPool(tickets, winners, identifier) {
  const drawn = new Set(winners.map((winner) => String(winner.record?.[identifier] ?? '').toLowerCase()));
  return tickets.filter((ticket) => !drawn.has(String(ticket[identifier] || '').toLowerCase()));
}

/**
 * The prize position a given draw awards.
 *
 * Draws happen in sequence, but the rank they carry depends on which end of
 * the list the organiser chose to start from: highest-first awards 1st, 2nd,
 * 3rd in turn, lowest-first counts back from the last prize so the evening
 * builds to the top one.
 */
function prizeNumberForDraw(appSettings, drawIndex) {
  const total = appSettings.totalPrizes;
  if (appSettings.prizes.drawOrder === 'lowest-first') {
    return Math.max(1, total - drawIndex);
  }
  return drawIndex + 1;
}

function buildStats(appSettings, tickets, winners) {
  const totalPrizes = appSettings.totalPrizes;
  const remainingTickets = remainingPool(tickets, winners, appSettings.data.identifier).length;

  return {
    totalPrizes,
    winnersCount: winners.length,
    remainingPrizes: Math.max(0, totalPrizes - winners.length),
    totalTickets: tickets.length,
    remainingTickets,
    isComplete: winners.length >= totalPrizes || remainingTickets === 0,
    // What the next Start will award, so the board can announce it first.
    nextPrizeNumber: winners.length < totalPrizes ? prizeNumberForDraw(appSettings, winners.length) : null,
  };
}

/**
 * Performs one draw and persists it. Returns a discriminated result rather
 * than throwing, because "no prizes left" is an expected outcome on stage.
 */
function drawWinner(appSettings) {
  const { tickets } = readTickets(appSettings);
  const state = loadDrawState();

  if (tickets.length === 0) {
    return { ok: false, reason: 'no-tickets', message: 'No participants have been uploaded yet.' };
  }

  if (state.winners.length >= appSettings.totalPrizes) {
    return { ok: false, reason: 'prizes-exhausted', message: 'All prizes have already been awarded.' };
  }

  const pool = remainingPool(tickets, state.winners, appSettings.data.identifier);
  if (pool.length === 0) {
    return { ok: false, reason: 'pool-empty', message: 'Every ticket has already been drawn.' };
  }

  // crypto.randomInt is uniform; Math.random is not suitable for a draw whose
  // fairness has to be defensible.
  const selected = pool[crypto.randomInt(0, pool.length)];
  const winner = {
    // The position awarded, which is not the sequence when the organiser
    // draws from the lowest prize upwards.
    prizeNumber: prizeNumberForDraw(appSettings, state.winners.length),
    drawIndex: state.winners.length + 1,
    drawnAt: new Date().toISOString(),
    record: selected,
  };

  const nextState = {
    ...state,
    startedAt: state.startedAt || winner.drawnAt,
    winners: [...state.winners, winner],
  };

  saveDrawState(nextState);

  return { ok: true, winner, stats: buildStats(appSettings, tickets, nextState.winners) };
}

function undoLastWinner(appSettings) {
  const state = loadDrawState();
  if (state.winners.length === 0) {
    return { ok: false, reason: 'nothing-to-undo', message: 'There are no draws to undo.' };
  }

  const removed = state.winners[state.winners.length - 1];
  const nextState = { ...state, winners: state.winners.slice(0, -1) };
  saveDrawState(nextState);

  return { ok: true, removed, stats: buildStats(appSettings, readTickets(appSettings).tickets, nextState.winners) };
}

function resetDraw(appSettings) {
  saveDrawState({ winners: [], startedAt: null });
  return { ok: true, stats: buildStats(appSettings, readTickets(appSettings).tickets, []) };
}

module.exports = {
  prizeNumberForDraw,
  loadDrawState,
  saveDrawState,
  readTickets,
  remainingPool,
  buildStats,
  drawWinner,
  undoLastWinner,
  resetDraw,
};
